gtsfm/bundle/bundle_adjustment.py

```diff
@@ -5,37 +5,147 @@
+    NonlinearFactorGraph,
+    PinholeCameraCal3Bundler,
+    PriorFactorCal3Bundler,
+    PriorFactorPose3,
+    SfmTrack,
+    Values,
+    symbol_shorthand,
+)
+from gtsam.noiseModel import Isotropic
 
 import gtsfm.utils.logger as logger_utils
 from gtsfm.common.gtsfm_data import GtsfmData
 from gtsfm.evaluation.metrics import GtsfmMetric, GtsfmMetricsGroup
 
 METRICS_PATH = Path(__file__).resolve().parent.parent.parent / "result_metrics"
 
+"""In this file, we use the GTSAM's GeneralSFMFactor2 instead of GeneralSFMFactor because Factor2 enables decoupling
+of the camera pose and the camera intrinsics, and hence gives an option to share the intrinsics between cameras.
+"""
+
 # TODO: any way this goes away?
-C = symbol_shorthand.C
-P = symbol_shorthand.P
+P = symbol_shorthand.P  # 3d point
+X = symbol_shorthand.X  # camera pose
+K = symbol_shorthand.K  # calibration
 
-PINHOLE_CAM_CAL3BUNDLER_DOF = 9  # 6 dof for pose, and 3 dof for f, k1, k2
+CAM_POSE3_DOF = 6  # 6 dof for pose of camera
+CAM_CAL3BUNDLER_DOF = 3  # 3 dof for f, k1, k2 for intrinsics of camera
 IMG_MEASUREMENT_DIM = 2  # 2d measurements (u,v) have 2 dof
 POINT3_DOF = 3  # 3d points have 3 dof
 
+
+# noise model params
+CAM_POSE3_PRIOR_NOISE_SIGMA = 0.1
+CAM_CAL3BUNDLER_PRIOR_NOISE_SIGMA = 0.1
+MEASUREMENT_NOISE_SIGMA = 1.0  # in pixels
+
 logger = logger_utils.get_logger()
 
 
-class BundleAdjustmentOptimizer(NamedTuple):
+class BundleAdjustmentOptimizer:
     """Bundle adjustment using factor-graphs in GTSAM.
 
     This class refines global pose estimates and intrinsics of cameras, and also refines 3D point cloud structure given
     tracks from triangulation."""
 
-    output_reproj_error_thresh: float
+    def __init__(self, output_reproj_error_thresh: float, shared_calib: bool = False):
+        """Initializes the parameters for bundle adjustment module.
+
+        Args:
+            output_reproj_error_thresh: the max reprojection error allowed in output.
+            shared_calib (optional): Flag to enable shared calibration across
+                                     all cameras. Defaults to False.
+        """
+        self._output_reproj_error_thresh = output_reproj_error_thresh
+        self._shared_calib = shared_calib
+
+    def __add_camera_prior(
+        self,
+        graph: NonlinearFactorGraph,
+        camera: PinholeCameraCal3Bundler,
+        camera_idx: int,
+    ) -> None:
+        """Add a prior factor in the factor graph for the camera parameters.
+
+        Args:
+            graph: factor graph for the problem.
+            camera: the camera object to add to graph and to initial values.
+            camera_idx: index of the camera.
+        """
+        graph.push_back(
+            PriorFactorPose3(X(camera_idx), camera.pose(), Isotropic.Sigma(CAM_POSE3_DOF, CAM_POSE3_PRIOR_NOISE_SIGMA))
+        )
+
+        # process just one camera if shared calibration; otherwise process all
+        if camera_idx == 0 or not self._shared_calib:
```

Comment:
maybe needs a flag to tell if a calibration prior has been added instead of check on index.

Reply:
i think this check should come on the caller side, not inside the subroutine. otherwise it can confuse people, thinking we add a prior for every camera @ayushbaid